hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -1,21 +1,63 @@
 //! The Yk PT trace decoder.
+//!
+//! The decoder works in two modes:
+//!
+//!  - compiler-assisted decoding
+//!  - disassembly-based decoding
+//!
+//! The former mode is for decoding portions of the trace that are for "native code" (code compiled
+//! by ykllvm). Code built with ykllvm gets static control flow information embedded in the end
+//! binary in a special section. Along with dynamic information provided by the PT packet stream,
+//! we have everything we need to decode these parts of the trace without disassembling
+//! instructions. This is the preferred mode of decoding.
+//!
+//! The latter mode is a fallback for decoding portions of the trace that are for "foreign code"
+//! (code not built with ykllvm). For foreign code there is no static control flow edge information
+//! available, so to decode these parts of the trace we have to disassemble the instruction stream
+//! (like libipt does).
+//!
+//! You may now be asking: why not just skip the parts of the trace that are for foreign code?
+//! After all, if a portion of code wasn't built with ykllvm, then we won't have IR for it anyway,
+//! meaning that the JIT is unable to inline it and we'd have to emit a call into the JIT trace.
+//! Why bother decoding the bit of the trace we don't actually care about?
+//!
+//! The problem is, it's not always easy to identify which parts of a PT trace are for native or
+//! foreign code: it's easy if the CPU that doesn't implement the deferred TIP optimisation (see
+//! the Intel 64 and IA32 Architectures Software Developer's Manual, Vol 3, Section  32.4.2.3).
+//!
+//! Deferred TIPs mean that TNT decisions can come in "out of order" with TIP updates. For example,
+//! a packet stream `[TNT(0,1), TIP(addr), TNT(1,0)]` may be optimised to `[TNT(0, 1, 1, 0),
+//! TIP(addr)]`. If the TIP update to `addr` is the return from foreign code, then when we resume
+//! compiler-assisted decoding then we need only the last two TNT decisions in the buffer
+//! (discarding the first two as we skip over foreign code). The problem is that without successor
+//! block information about the foreign code we can't know how how many TNT decisions correspond
+//! with the foreign code, and thus how many decisions to discard.
+//!
+//! We therefore surrender and disassemble foreign code, popping TNT decisions as we encounter
```

Comment:
s/surrender and/have to/ ?

Reply:
Fixed in d4bdb709ef7e773e42409937f0f7d1e6452ff43e